### PR TITLE
fix: ensure pusher events are processed in the same order they were received

### PR DESCRIPTION
**Pull request by barros001**
### Details

This PR will ensure that all pusher events are processed in the exact same order as they were received.

### Fixed Issues

$ https://github.com/Expensify/App/issues/23679
PROPOSAL: https://github.com/Expensify/App/issues/23679#issuecomment-1793464745

### Tests

1. Login as user A and open a report
2. On a different browser (or incognito window), login as user B and open the same report
3. As user A, send a message
4. Disable network connection as to make both accounts offline
5. As user A, react to the previous message with a 👍
6. As user B, react to the previous message with 👍 
7. As user A, remove the reaction
8. As user B, remove the reaction
9. Reactivate the network connection
10. Verify that after network sync, no reactions are left on the message for both users

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Login as user A and open a report
2. On a different browser (or incognito window), login as user B and open the same report
3. As user A, send a message
4. Disable network connection as to make both accounts offline
5. As user A, react to the previous message with a 👍
6. As user B, react to the previous message with 👍 
7. As user A, remove the reaction
8. As user B, remove the reaction
9. Reactivate the network connection
10. Verify that after network sync, no reactions are left on the message for both users

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/765936/36edf3f4-de2c-41b9-820a-42c9fddc0bef

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/765936/ba0885d8-d75c-43b1-a336-de2cec77c601

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/765936/d63ae70c-e657-400d-b593-d4bcf09e3248

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/765936/8cc1f607-0ae5-4d39-8b4e-8aa397577b08

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/765936/40171855-f4fb-48f8-8cf2-94867f4151f2

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/765936/d75dbae8-21a1-454f-9438-fbf015b669dc

</details>

### CODE REVIEW

File: src/libs/actions/OnyxUpdates.ts [hunk 1]
```diff
@@ -44,11 +48,17 @@ function applyHTTPSOnyxUpdates(request: Request, response: Response) {
...
+        console.debug('[OnyxUpdateManager] Applying pusher update');
     });
+
+    pusherEventsPromise = updates
+        .reduce((promise, update) => promise.then(() => PusherUtils.triggerMultiEventHandler(update.eventType, update.data)), pusherEventsPromise)
+        .then(() => {
+            console.debug('[OnyxUpdateManager] Done applying Pusher update');
+        });
+
+    return pusherEventsPromise;
```
[the remark this answers is not in the document]
Comment (barros001) [reply src/libs/actions/OnyxUpdates.ts:61]: Agreed. Originally I went this route but then lint yelled at me. I also had a for loop initially that might be more readable, but ended up going with a reduce. I can switch back to it if that makes things easier to understand.

**Comment by barros001** [comment]
> @barros001 I still have the case when issue still happens. Please try to follow step in my video exactly
Ok, will take a look tomorrow.

**Comment by barros001** [comment]
@aimane-chnaif I tried to reproduce the issue several times but I could not reproduce it. Is this something you can reliably reproduce? I noticed you had a 3rd instance of the app running on a what appeared to be an iOS simulator. Can you give me more details about your test setup? There clearly is something going on as shown on your video, but so far I have not been able to reproduce.

**Comment by barros001** [comment]
@aimane-chnaif I believe we still have a race condition here and my solution solved only part of it. See code below:

```ts
function apply({lastUpdateID, type, request, response, updates}: OnyxUpdatesFromServer): Promise<void | Response> | undefined {
    console.debug(`[OnyxUpdateManager] Applying update type: ${type} with lastUpdateID: ${lastUpdateID}`, {request, response, updates});

    if (lastUpdateID && lastUpdateIDAppliedToClient && Number(lastUpdateID) < lastUpdateIDAppliedToClient) {
        console.debug('[OnyxUpdateManager] Update received was older than current state, returning without applying the updates');
        return Promise.resolve();
    }
    if (lastUpdateID && (lastUpdateIDAppliedToClient === null || Number(lastUpdateID) > lastUpdateIDAppliedToClient)) {
        Onyx.merge(ONYXKEYS.ONYX_UPDATES_LAST_UPDATE_ID_APPLIED_TO_CLIENT, Number(lastUpdateID));
    }
    if (type === CONST.ONYX_UPDATE_TYPES.HTTPS && request && response) {
        return applyHTTPSOnyxUpdates(request, response);
    }
    if (type === CONST.ONYX_UPDATE_TYPES.PUSHER && updates) {
        return applyPusherOnyxUpdates(updates);
    }
}
```

This gets called every time a push event is received, but it also gets called when http events are received. While debugging I noticed that sometimes a push even start processing and before it ends, http events come in and process first. While I could not reproduce the issue you showed above, there's a good chance that this could lead to inconsistent state.

This makes me think that, although we still need to serialize the push events inside `applyPusherOnyxUpdates`, we also need to serialize calls to `apply` so multiple calls to it don't get processed out of order.

Also, `Onyx.merge(ONYXKEYS.ONYX_UPDATES_LAST_UPDATE_ID_APPLIED_TO_CLIENT, Number(lastUpdateID));` also returns a promise that's not waited on, so things could happen while we're waiting for the data to be persisted. 

The resulting code would look something like this:

```ts
let applyPromise = Promise.resolve();
function apply({lastUpdateID, type, request, response, updates}: OnyxUpdatesFromServer): Promise<void | Response> | undefined {
    // ensure paralell calls to this method will execute sequentially
    applyPromise = applyPromise.then(() => {
        console.debug(`[OnyxUpdateManager] Applying update type: ${type} with lastUpdateID: ${lastUpdateID}`, {request, response, updates});

        if (lastUpdateID && lastUpdateIDAppliedToClient && Number(lastUpdateID) < lastUpdateIDAppliedToClient) {
            console.debug('[OnyxUpdateManager] Update received was older than current state, returning without applying the updates');
            return Promise.resolve();
        }

        let onyxMergePromise = Promise.resolve();

        if (lastUpdateID && (lastUpdateIDAppliedToClient === null || Number(lastUpdateID) > lastUpdateIDAppliedToClient)) {
            onyxMergePromise = Onyx.merge(ONYXKEYS.ONYX_UPDATES_LAST_UPDATE_ID_APPLIED_TO_CLIENT, Number(lastUpdateID));
        }

        return onyxMergePromise.then(() => {
            if (type === CONST.ONYX_UPDATE_TYPES.HTTPS && request && response) {
                return applyHTTPSOnyxUpdates(request, response);
            }
            if (type === CONST.ONYX_UPDATE_TYPES.PUSHER && updates) {
                return applyPusherOnyxUpdates(updates);
            }
        });
    });

    return applyPromise;
```

Code still needs a little cleanup and more tests but before I go to the work of implementing this I wanted to get some feedback. The code above combined with what's already in the PR will ensure every event received is processed in the correct order.

**Comment by barros001** [comment]
Let me step back a little first, I'll try to better explain what's going on here. If I refresh the app, or go back online after being offline for a while, the app will do a bunch of things such as pull data from the BE and apply to the FE. Here's some sequence of logs I see sometimes (not always, but usually going from offline to online).

Screenshot 2023-11-14 at 12 05 34 PM

Do you see anything funny?

* [OnyxUpdateManager] Applying update type: pusher with lastUpdateID: 34773552
* ...
* [OnyxUpdateManager] Applying pusher update
* [OnyxUpdateManager] Applying update type: https with lastUpdateID: 0 
* [OnyxUpdateManager] Applying https update
* [OnyxUpdateManager] Done applying HTTPS update
* ....
* ....
* [OnyxUpdateManager] Done applying Pusher update

An HTTP update came in while we were applying a pusher update and executed all the way through before the pusher event was fully processed. I don't think this is how this was intended as this processes things out of sequence.

`apply` gets called from the `SaveResponseInOnyx` middleware:

https://github.com/Expensify/App/blob/183cce04c94084ec76ff3ada90fe2c6ef4ad20d6/src/libs/Middleware/SaveResponseInOnyx.ts#L39-L41

And also from here:

https://github.com/Expensify/App/blob/ca13914a7b4249585f178b396f279c7657f48f40/src/libs/actions/User.js#L514-L517

These two places are unaware of each other causing them to call apply in parallel and events (http/pusher) being mixed up and potentially executed out of order.

The PR so far fixes the initial side effect but there seems to be a more fundamental issue going on here and apply needs to be serialized as a whole to avoid potential issues. I'll dig a little deeper to see if I can find a good solution for this. I tried applying the concept I described in the previous comment above but I think there's some sort of deadlock happening and FE gets stuck waiting to finish processing a push event, which never happens.

Answering your question about using async/await, I don't think I can really convert it to use async/await due to the nature of the problem: single function, multiple entrypoints where both entry points must wait any pending promise before executing the next one.

I'll dig a little further and see if I can figure a clean solution.

As far as the issue you pointed out, are you still able to reproduce it? Does it happen all the time? I could not reproduce it yet on my end.

**Comment by barros001** [comment]
Interesting.. If I'm on main, I can reliably reproduce it using the exact same steps are you, but if I'm on the branch where the fix is applied, I can no longer replicate, the issue seems to be resolved. I'll try a few more time with different accounts.

Silly question, you don't happen to be on `main` by accident, do you? 😄

**Comment by barros001** [comment]
@aimane-chnaif there seems to be something else going on on your video. Slow it down and watch it closely. 

* You add a black TU
* Then a yellow TU
* Then remove black TU
* Finally remove the yellow black TU.

You then go back to and make the first window (that sent a black TU) online, once you switch back to the other (that sent the yellow TU), which is still offline, I see the black TU there, shaded as if that window tried to send it while still offline, And it all happens before you go back online. See below:

Screenshot 2023-11-14 at 1 36 33 PM

Note the Offline message below composer, top right corner still says offline, and black TU is there. I'm wondering where that came from if that account was still offline.

**Comment by barros001** [comment]
I just pushed a new commit addressing my [previous comment](https://github.com/Expensify/App/pull/31144#issuecomment-1810798283). `applyPusherOnyxUpdates` had an issue with multiple event being processed in parallel, which the new update also addresses, but IMO the real issue is `apply` being called in parallel.

Can you give it a try and see if you still reproduce it?

**Comment by barros001** [comment]
sorry about that, I did notice the same, there seem to be some sort of deadlock somewhere.. Reverted and we're back to the original solution. I really wish I could reproduce what you're seeing.

**Comment by tgolen** [comment]
Sure, I've triggered one for you. FYI, any internal engineer can trigger those, so no need to ping me directly for them.

**Comment by tgolen** [comment]
Oh, if I am on the linked issue, then that's totally fine. Apologies! I was
not making that connection and I only looked to see if I had previously
been involved in the PR.

**Review by aimane-chnaif — APPROVED** [review]

**Review by tgolen — APPROVED** [review]

**Comment by nyomanjyotisa** [comment]
The shared Pusher promise chain introduced in #31144 lacked rejection recovery, causing #100174: one failed apply prevented subsequent Pusher updates from running for the rest of the session. Fixed in #100594 with chain recovery and watermark continuity protection.

**Pull request merged**